state_machine/applications/flight/Tasks/radio.py

```diff
@@ -105,24 +52,88 @@ async def main_task(self):
 
             if tq.peek().done():
                 tq.pop()
+        else:
+            self.debug("No packets to send")
+            cubesat.radio.listen()
+            response = await cubesat.radio.receive(keep_listening=True, with_ack=ANTENNA_ATTACHED, timeout=10)
+            if response is not None:
+                header = response[0]
+                response = response[1:]  # remove the header byte
+
+                self.debug(f'Recieved msg "{response}", RSSI: {cubesat.radio.last_rssi - 137}')
+
+                if header == headers.NAIVE_START or header == headers.NAIVE_MID or header == headers.NAIVE_END:
+                    self.handle_naive(header, response)
+                elif header == headers.CHUNK_START or header == headers.CHUNK_MID or header == headers.CHUNK_END:
+                    self.handle_chunk(header, response)
+                elif header == headers.COMMAND:
+                    self.handle_command(response)
+            else:
+                self.debug('No packets received')
+
         cubesat.radio.sleep()
 
     def handle_naive(self, header, response):
         if header == headers.NAIVE_START:
-            txt = str(response, 'ascii')
-            self.msg = txt
-            self.last = txt
-            print('Started recieving message')
-        elif header == headers.NAIVE_MID:
-            txt = str(response, 'ascii')
-            if txt == self.last:
-                print('Repeated message')
+            self.msg_last = response
+            self.msg = response
+        else:
+            if response != self.msg_last:
+                self.msg += response
+            else:
+                self.debug('Repeated chunk')
+
+        if header == headers.NAIVE_END:
+            self.cmsg_last = None
+            self.msg = str(self.msg, 'utf-8')
+
+    def handle_command(self, response):
+        if len(response) < 6 or response[:4] != self.super_secret_code:
+            return
+
+        cmd = bytes(response[4:6])  # [pass-code(4 bytes)] [cmd 2 bytes] [args]
+        cmd_args = bytes(response[6:])
+        try:
+            self.debug(f'cmd args: {cmd_args}', 2)
+        except Exception as e:
+            self.debug(f'arg decoding error: {e}', 2)
+
+        if cmd in cdh.commands:
+            try:
+                if cmd_args:
+                    self.debug(f'running {cdh.commands[cmd]} (with args: {cmd_args})')
+                    cdh.commands[cmd](self, cmd_args)
+                else:
+                    self.debug(f'running {cdh.commands[cmd]} (no args)')
+                    cdh.commands[cmd](self)
+            except Exception as e:
+                self.debug(f'something went wrong: {e}')
+                cubesat.radio.send(str(e).encode())
+        else:
+            self.debug('invalid command!')
+            cubesat.radio.send(b'invalid cmd' + cmd)
+
+    def handle_chunk(self, header, response):
```

Comment:
I think some more documentation on this function and the one above would help; I believe handle_command was part of cdh, so it may not have been previously documented, but I would still suggest adding comments.